### PR TITLE
enable CloudTrail action

### PR DESCRIPTION
**Pull request by skurfer**
Another stab at #339. This will enable the trail(s) named in the policy, or create then enable them (and a dedicated bucket) if they don’t exist.

### CODE REVIEW

**Comment by kapilt** [comment]
actions now require a permissions tuple describing the permissions they need.

File: c7n/resources/cloudtrail.py [hunk 1]
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
...
+            account_id = ''
+        for trail_name in to_create:
+            # create a trail (and its bucket)
+            bucket_name = '{}-trails'.format(trail_name)
+            s3client = session.client('s3')
+            s3client.create_bucket(Bucket=bucket_name)
+            policy_json = policy_json_tmpl.format(
+                bucket_name=bucket_name, account_id=account_id,
+            )
+            s3client.put_bucket_policy(Bucket=bucket_name, Policy=policy_json)
```
Comment (kapilt) [inline c7n/resources/cloudtrail.py]: this needs to play nice with an existing bucket policy

File: c7n/resources/cloudtrail.py [hunk 2]
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
...
+            new_trail = client.create_trail(
+                Name=trail_name,
+                S3BucketName=bucket_name,
```
Comment (kapilt) [inline c7n/resources/cloudtrail.py]: the other options here need to be exposed for config or defaulted (global events, multi-region, kms key, etc).

File: c7n/resources/cloudtrail.py [hunk 3]
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
...
+          "s3:x-amz-acl": "bucket-owner-full-control"
+        }}
+      }}
+    }}
+  ]
+}}'''
+
+
+@actions.register('enable')
+class EnableTrail(BaseAction):
```
Comment (kapilt) [inline c7n/resources/cloudtrail.py]: actions and filters need iam permissions as class attribute

File: c7n/resources/cloudtrail.py (hunk 1 above)
Comment (skurfer) [inline c7n/resources/cloudtrail.py]: The bucket was created one line earlier, so existing policy shouldn’t be a concern. Is it possible to define a default policy for newly created buckets? That’s the only way I could see it already having one.

Comment (kapilt) [inline c7n/resources/cloudtrail.py]: thats fair, but the consideration is that we'll be doing this to trails back into place if their deleted maliciously, in which case the bucket may already exist and we're just re-establishing the connection and the bucket may already exist (aka try/except on bucket exists).

Comment (skurfer) [inline c7n/resources/cloudtrail.py]: If a trail is deleted, it won’t be passed to the `process()` method, so we won’t have any metadata about it (like which bucket it was attached to). The bucket here is completely made up based on the name of the trail. (I suppose a collision is still possible, so we can test for existing.)

Should we allow (require?) a bucket name for each trail in the policy so the code doesn’t have to guess?

Comment (kapilt) [inline c7n/resources/cloudtrail.py]: good point.. so the end goal is to have something like 'ensure' trail, such that no matter what an active trail is running, that we can trigger off events that modify or disable trails or for initial account setup.

**Comment by skurfer** [comment]
This has been updated to support the various options.

Like I mentioned on the call, the coverage didn’t match my expectations. I ran through in the debugger and at least two of the lines that supposedly aren’t getting tested *are* getting called when tests run, so I’m chalking it up to weirdness in the coverage package and submitting this as is.

**Comment by skurfer** [comment]
Updates based on the last discussion:

* Moved the enable trail action to the account resource
* Added an explicit call to `describe_trails()` since we’re now being passed a list of accounts instead of trails
* Added DescribeTrails to the list of permissions
* moved and updated tests

**Comment by kapilt** [comment]
branch updated and pull request resubmitted in #1098

**Pull request closed without merging**